docs: add ivy opt-in docs (#28569)
angular/angular: aio/content/guide/ivy-preview.md

> # Previewing Ivy with Angular CLI

Reviewer: ```suggestion
# Opt-ing into Angular Ivy with Angular CLI
```

> Starting with Angular version 8, you can experimentally try out the new [Ivy Compiler](link to some ivy announcement here).

Reviewer: ```suggestion
Starting with Angular version 8, you can experimentally try out the new [Ivy Compiler](https://blog.angular.io/a-plan-for-version-8-0-and-ivy-b3318dfc19f7).
```

## Starting a new project using Ivy

> To start a new project with Ivy enabled, use the `--experimental-ivy` flag with the [`ng new`](cli/new) command:

Reviewer: did we notice that the flag name is different between CLI and tsconfig? "experimental-ivy" vs "enableIvy" - could we consolidate these?

Reviewer: @alexeagle @kara @alxhub @filipesilva ⬆️

Author: I think the easiest way to consolidate is to change the CLI `--experimental-ivy` to `--enable-ivy`. 

We could do that, but then we lose the implicit opt-in that goes with a user having to type out the `experimental` word. But that's ok if we commit to keep the `--enable-ivy` flag around and it not being experimental at all.

```sh
ng new shiny-ivy-app --experimental-ivy
```

Everything will be configured for you:
- `enableIvy` set to `true` in `src/tsconfig.app.json`.
- `"aot": true` added to your default build options.
- `postinstall` script for the [Angular Compatibility Compiler](#ngcc).

{@a updating}
## Updating an existing project to use Ivy

By configuring a few key files, you can also update your existing project(s) to use Ivy.

- Add the `enableIvy` option in the `angularCompilerOptions` in your project's `src/tsconfig.app.json`:
```json
{
  "compilerOptions": { ... },
  "angularCompilerOptions": {
    "enableIvy": true,

>     "allowEmptyCodegenFiles": true,

Reviewer: ```suggestion
    # you might also need to enable shim file generation
    # this requirement should go away before version 8 ships
    "allowEmptyCodegenFiles": true,
```

Author: Although `tsconfig.json` itself accepts comments, the `src/tsconfig.app.json` file is not automatically identified by VS Code as being a TypeScript configuration file. This means comments appear as an error.

Instead of adding this comment, I added text to the instruction above the code.

>     "allowEmptyCodegenFiles": true,

Reviewer: @kara @alxhub  we should create a flag with a human understandable name that is related to shim generation instead of piggybacking on this one, e.g. "generateNgFactoryShimFiles" or similar.

  }
}
```
- Set the default build options for your project to always use AOT compilation if it isn't already:
```json
{
  "projects": {
    "my-existing-project": {
      "architect": {
        "build": {
          "options": {
            ...
            "aot": true,
          }
        }
      }
    }
  }
}
```
- Add a `postinstall` script to the workspace `package.json` file to run the [Angular Compatibility Compiler](#ngcc):
```json
{
  "scripts": {
    ...

>     "postinstall": "ivy-ngcc"

Reviewer: @petebacondarwin @gkalpak 
@kara @alxhub  `ivy-ngcc` is an odd looking binary filename. why not just call it `ngcc`?

Reviewer 2: +1 on `ngcc`

Reviewer 3: > why not just call it `ngcc`?

Maybe because the next time we need a compatibility compiler, we would have to name it `ngcc2` (a la `Renderer2`), which is not cool 😛

  }
}
```

> - Reinstall your `package.json` dependencies to run the newly added script.

Reviewer: ````suggestion
- Reinstall your `package.json` dependencies to run the newly added script:

```
npm install
```
````

## Switching back to the current compiler

To stop using the Ivy compiler you need to undo the steps taken when [updating to use Ivy](#updating).
- Set `enableIvy` to false in `src/tsconfig.app.json`, or remove it completely.
- Remove `"aot": true` from your default build options if you didn't have it there before.
- Remove the `postinstall` script.
- Delete and reinstall your node modules.

{@a ngcc}
## The Angular Compatibility Compiler

Your third party libraries also need to be compiled with Ivy for you to use them in your Ivy application.
That's where the Angular Compatibility (`ngcc`) compiler comes into play.

We set it as a postinstall script in your `package.json` so that it always runs when you install
your `package.json` dependencies.

Using a postinstall hook to run `ngcc` is just a temporary integration. 
We expect `ngcc` to be seamlessly integrated into the Angular CLI build pipeline in the future before the full Ivy rollout. 
Once that's implemented `ngcc` will and not be visible to developers.

> Until that happens, opting into Ivy means that all projects in a single CLI workspace will be compiled with Ivy.

Reviewer: ```suggestion
Until that happens, opting into Ivy means that all projects in a single CLI workspace will be compiled with Ivy.

Note: we are aware of build-time performance issues with ngcc. Please bear with us while we improve the tooling and build integration in the Angular CLI.
```
